File: testdata/fixtures/go/pkg/service.go
// Package pkg provides HTTP handling and business logic.
package pkg

import "fixture/internal"

// Service defines the business logic interface.
type Service interface {
	Process(input string) string
	Validate(input string) error
}

// DefaultService is the default implementation of Service.
type DefaultService struct {
	model *Model
}

// NewDefaultService creates a new DefaultService.
func NewDefaultService() *DefaultService {
	return &DefaultService{
		model: NewModel("default"),
	}
}

// Process implements Service.Process.
// Call chain: Process -> model.Transform -> internal.FormatOutput
func (s *DefaultService) Process(input string) string {
	transformed := s.model.Transform(input)
	return internal.FormatOutput(transformed)
}

// Validate implements Service.Validate.
func (s *DefaultService) Validate(input string) error {
	if input == "" {
		return internal.ErrEmptyInput
	}
	return nil
}

// CachingService wraps a Service with caching.
type CachingService struct {
	inner Service
	cache map[string]string
}

// NewCachingService creates a caching wrapper around a Service.
func NewCachingService(inner Service) *CachingService {
	return &CachingService{
		inner: inner,
		cache: make(map[string]string),
	}
}

// Process implements Service.Process with caching.
func (c *CachingService) Process(input string) string {
	if cached, ok := c.cache[input]; ok {
		return cached
	}
	result := c.inner.Process(input)
	c.cache[input] = result
	return result
}

// Validate implements Service.Validate by delegating.
func (c *CachingService) Validate(input string) error {
	return c.inner.Validate(input)
}

File: testdata/fixtures/typescript/src/pkg/service.ts
/**
 * Service module - business logic with interface pattern
 */

import { Model } from "./model";
import { ValidateInput, SanitizeInput } from "../internal/util";

/**
 * Service interface defining business operations
 */
export interface Service {
  /**
   * Validates the input before processing
   * @param input The input to validate
   * @throws Error if validation fails
   */
  validate(input: string): void;

  /**
   * Processes input and returns result
   * @param input The input to process
   * @returns The processed result
   */
  process(input: string): string;
}

/**
 * Default service implementation
 */
export class DefaultService implements Service {
  private model: Model;

  constructor(model: Model) {
    this.model = model;
  }

  validate(input: string): void {
    ValidateInput(input);
  }

  process(input: string): string {
    this.validate(input);
    const clean = SanitizeInput(input);
    return this.model.transform() + ": " + clean;
  }
}

/**
 * Factory function for creating default service
 * @param model The model to use
 * @returns New DefaultService instance
 */
export function newDefaultService(model: Model): DefaultService {
  return new DefaultService(model);
}

/**
 * Caching service implementation with memoization
 */
export class CachingService implements Service {
  private inner: Service;
  private cache: Map<string, string>;

  constructor(inner: Service) {
    this.inner = inner;
    this.cache = new Map();
  }

  validate(input: string): void {
    this.inner.validate(input);
  }

  process(input: string): string {
    const cached = this.cache.get(input);
    if (cached !== undefined) {
      return cached;
    }
    const result = this.inner.process(input);
    this.cache.set(input, result);
    return result;
  }
}

/**
 * Factory function for creating caching service
 * @param inner The service to wrap
 * @returns New CachingService instance
 */
export function newCachingService(inner: Service): CachingService {
  return new CachingService(inner);
}
